internal/handlers/pg/pgdb/query.go

```diff
@@ -268,61 +267,133 @@ func prepareWhereClause(sqlFilters *types.Document) (string, []any, error) {
-				filters = append(filters, fmt.Sprintf(sql, p.Next(), p.Next()))
-				args = append(args, k, string(must.NotFail(pjson.MarshalSingleValue(v))))
-
-			default:
-				panic(fmt.Sprintf("Unexpected type of value: %v", v))
+		switch v := v.(type) {
+		case *types.Document:
+			iter := v.Iterator()
+			defer iter.Close()
+
+			for {
+				docKey, docVal, err := iter.Next()
+				if err != nil {
+					if errors.Is(err, iterator.ErrIteratorDone) {
+						break
+					}
+
+					return "", nil, lazyerrors.Error(err)
+				}
+
+				switch docKey {
+				case "$eq":
+					switch docVal := docVal.(type) {
+					case *types.Document, *types.Array, types.Binary, bool,
+						time.Time, types.NullType, types.Regex, types.Timestamp:
+						// type not supported for pushdown
+					case float64, string, types.ObjectID, int32, int64:
+						builder.addFilter(k, eq, docVal)
+					default:
+						panic(fmt.Sprintf("Unexpected type of value: %v", v))
+					}
+
+				case "$ne":
+					switch docVal := docVal.(type) {
+					case *types.Document, *types.Array, types.Binary, bool,
+						time.Time, types.NullType, types.Regex, types.Timestamp:
+						// type not supported for pushdown
+					case float64, string, types.ObjectID, int32, int64:
+						builder.addFilter(k, ne, docVal)
+					default:
+						panic(fmt.Sprintf("Unexpected type of value: %v", v))
+					}
+
+				default:
+					// TODO $gt and $lt https://github.com/FerretDB/FerretDB/issues/1875
+					continue
+				}
 			}
 
-			continue
-		}
-
-		switch v := v.(type) {
-		case *types.Document, *types.Array, types.Binary, bool, time.Time, types.NullType, types.Regex, types.Timestamp:
+		case *types.Array, types.Binary, bool, time.Time, types.NullType, types.Regex, types.Timestamp:
 			// type not supported for pushdown
-			// TODO $eq and $ne https://github.com/FerretDB/FerretDB/issues/1840
-			// TODO $gt and $lt https://github.com/FerretDB/FerretDB/issues/1875
 			continue
 
 		case float64, string, types.ObjectID, int32, int64:
-			// Select if value under the key is equal to provided value.
-			// If the value under the key is not equal to v,
-			// but the value under the key k is an array - select if it contains the value equal to v.
-			sql := `(_jsonb->%[1]s)::jsonb = %[2]s OR (_jsonb->%[1]s)::jsonb @> %[2]s`
-
-			// placeholder p.Next() returns SQL argument references such as $1, $2 to prevent SQL injections.
-			// placeholder $1 is used for field key,
-			// placeholder $2 is used for field value v.
-			filters = append(filters, fmt.Sprintf(sql, p.Next(), p.Next()))
-			args = append(args, k, string(must.NotFail(pjson.MarshalSingleValue(v))))
+			builder.addFilter(k, eq, v)
 
 		default:
 			panic(fmt.Sprintf("Unexpected type of value: %v", v))
 		}
 	}
 
-	var query string
+	query, args := builder.generateWhereClause()
+
+	return query, args, nil
+}
+
+// builderOperator represents available filtersBuilder operators.
+type builderOperator uint8
+
+const (
+	eq builderOperator = iota
```

Comment:
> We have key like $eq and $ne.

In this example builderOperators `eq` and `ne` are not the same thing as `$eq` and `$ne`.
`$eq` is a FerretDB operator that handles different types in a specific way, it can contain much more FerretDB specific logic inside. 
On the other hand `eq` is representation of a simple PostgreSQL clause, which might be useful for `$eq` and other operators to work.

> Why do we need to introduce this?

It was an attempt to wrap some basic query building logic into the separate function. 
Currently for every operator we need to:
- check if the key is `_id` because we use different SQL operator for those
- check if the key is dot notation, because we use different SQL operator for those either
- write whole query for each operator
- append to the filters and args variables to generate actual SQL query afterwards

We need to do a lot of stuff just for the actual SQL query generation, but that's not all, 
as we also need to handle some FerretDB operators logic, such as handling different types in different way,
everything in a single function.

Imo it's good to move database level comparison into different place and make it easy to just call it with single lines of code. 
- It will prevent us from code repetition, as all the `_id`/dot notation logic can be handled inside of it.
- We wouldn't need to write more and more SQL code for every operator, but rather just "build" those queries with existing blocks
(equal operation is useful for many other cases than just `$eq`)
- `prepareWhereClause` function would just check for types and operator specific logic, and call appropriate builder functions
- It will also prevent us from introducing more and more of nested `switch`es, loops and `if`s.

And yes, as we agreed, we don't restructure this before GA, but afterwards I would strongly recommend to take a look at this, as
with re-introducing dot notation and adding a couple of additional operators, `prepareWhereClause` will be just painful to read...

Reply:
Thanks for the explanation. I think filter building of repetitiveness is actually fine. There are somethings that are easy to read when it is repetitive, I think sql is one of them. Because you are only focusing on one set of sql at a time. The nested switch is another story, yeah I agree it would be nice to prevent.